File: Problem Of The Day/Geeks for Geeks/Medium/AggressiveCows.cpp

// https://practice.geeksforgeeks.org/problems/aggressive-cows/1    

//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
public:

    int solve(int n, int k, vector<int> &stalls) {
    
        // Write your code here
        
        int low = 1;
        sort(stalls.begin(), stalls.end());
        int high = stalls.back() - stalls.front();
        int ans = 0;
        while(low <= high){
            int mid = low + (high - low)/2;
            int cnt = 1;
            int curr = stalls[0];
            for(int i = 1; i < n && cnt < k; i++){
                if(stalls[i] >= curr + mid){
                    cnt++;
                    curr = stalls[i];
                }
            }
            if(cnt < k){
                high = mid - 1;
            }
            else{
                ans = max(mid, ans);
                low = mid + 1;
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t = 1;
    cin >> t;

    // freopen ("output_gfg.txt", "w", stdout);

    while (t--) {
        // Input

        int n, k;
        cin >> n >> k;

        vector<int> stalls(n);
        for (int i = 0; i < n; i++) {
            cin >> stalls[i];
        }
        // char ch;
        // cin >> ch;

        Solution obj;
        cout << obj.solve(n, k, stalls) << endl;

        // cout << "~\n";
    }
    // fclose(stdout);

    return 0;
}
// } Driver Code Ends

File: DSA-Problems/BinarySearch/AggressiveCows.java

//https://www.codingninjas.com/codestudio/problems/aggressive-cows_1082559

import java.util.*;

class AggressiveCows 
{
    public static int aggressiveCows(ArrayList<Integer> stalls, int k) 
    {
        //    Write your code here.

        Collections.sort(stalls);

        int min=0;
        int max=-1;
        for(int i=0;i<stalls.size();i++){
            if(stalls.get(i)>max){
                max=stalls.get(i);
            }
        }
        int s=min, e=max, mid= s+(e-s)/2;
        int ans=0;
        while(s<=e){

            if(isPossibleSolution(stalls, k, mid)){
                ans=mid;
                s=mid+1;
            }else{
                e=mid-1;
            }
            mid=s+(e-s)/2;
        }
        return ans;
    }

    public static boolean isPossibleSolution(ArrayList<Integer> stalls, int k, int mid){

        int cowCount=1;
        int lastPosition=stalls.get(0);

        for(int i=0;i<stalls.size();i++){
            if(stalls.get(i)-lastPosition>=mid){
                lastPosition=stalls.get(i);
                cowCount++;
                if(cowCount==k){
                    return true;
                }   
            }
        }
        return false;        
    }
}